File: src/main/java/quebec/virtualite/backend/security/SecurityConfiguration.java
package quebec.virtualite.backend.security;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.config.annotation.method.configuration.EnableReactiveMethodSecurity;
import org.springframework.security.config.annotation.web.reactive.EnableWebFluxSecurity;
import org.springframework.security.config.web.server.ServerHttpSecurity;
import org.springframework.security.core.userdetails.MapReactiveUserDetailsService;
import org.springframework.security.core.userdetails.User;
import org.springframework.security.crypto.factory.PasswordEncoderFactories;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.web.server.SecurityWebFilterChain;

import static quebec.virtualite.backend.security.SecurityUsers.TEST_PASSWORD;
import static quebec.virtualite.backend.security.SecurityUsers.TEST_USER;

@Configuration
@EnableReactiveMethodSecurity
@EnableWebFluxSecurity
public class SecurityConfiguration
{
    @Bean
    public SecurityWebFilterChain springWebFilterChain(ServerHttpSecurity http)
    {
        return http
            .authorizeExchange()
            .anyExchange().authenticated()
            .and().httpBasic()
            .and().build();
    }

    @Bean
    public MapReactiveUserDetailsService userDetailsService()
    {
        return new MapReactiveUserDetailsService(
            User
                .withUsername(TEST_USER)
                .password(passwordEncoder().encode(TEST_PASSWORD))
                .roles("USER")
                .build());
    }

    private PasswordEncoder passwordEncoder()
    {
        return PasswordEncoderFactories.createDelegatingPasswordEncoder();
    }
}

File: src/main/java/quebec/virtualite/backend/security/SecurityConfiguration.kt
package quebec.virtualite.backend.security

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Configuration
import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter
import org.springframework.security.web.csrf.CsrfFilter
import org.springframework.security.web.csrf.CsrfToken
import org.springframework.security.web.csrf.CsrfTokenRepository
import org.springframework.security.web.csrf.HttpSessionCsrfTokenRepository
import org.springframework.web.filter.OncePerRequestFilter
import org.springframework.web.util.WebUtils
import java.io.IOException
import javax.servlet.FilterChain
import javax.servlet.ServletException
import javax.servlet.http.Cookie
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse
import javax.sql.DataSource

@Configuration
@EnableWebSecurity
open class SecurityConfiguration : WebSecurityConfigurerAdapter()
{
    @Autowired
    private lateinit var dataSource: DataSource

    @Autowired
    @Throws(Exception::class)

    fun configAuthentication(auth: AuthenticationManagerBuilder)
    {
        auth.jdbcAuthentication().dataSource(dataSource)
    }

    @Throws(Exception::class)
    override fun configure(http: HttpSecurity)
    {
        http.httpBasic()
            .and().authorizeRequests()
            .antMatchers("/css/**", "/i18n/**", "/js/**").permitAll()
            .antMatchers("/*.html", "/").permitAll()
            .anyRequest().authenticated()
            .and().formLogin().loginPage("/login").permitAll()
            .and().logout()
            .and().addFilterAfter(CsrfHeaderFilter(), CsrfFilter::class.java)
            .csrf().csrfTokenRepository(csrfTokenRepository())
    }

    private fun csrfTokenRepository(): CsrfTokenRepository
    {
        val repository = HttpSessionCsrfTokenRepository()
        repository.setHeaderName(X_XSRF_TOKEN)
        return repository
    }

    private class CsrfHeaderFilter : OncePerRequestFilter()
    {
        @Throws(ServletException::class, IOException::class)
        override fun doFilterInternal(request: HttpServletRequest,
                                      response: HttpServletResponse, filterChain: FilterChain)
        {
            val csrf = request.getAttribute(CsrfToken::class.java.name) as CsrfToken

            var cookie = WebUtils.getCookie(request, XSRF_TOKEN)
            val token = csrf.token

            if (cookie == null || token != null && token != cookie.value)
            {
                cookie = Cookie(XSRF_TOKEN, token)
                cookie.path = "/"
                response.addCookie(cookie)
            }

            filterChain.doFilter(request, response)
        }
    }

    companion object
    {
        private const val XSRF_TOKEN = "XSRF-TOKEN"
        private const val X_XSRF_TOKEN = "X-XSRF-TOKEN"
    }
}